pkg/printers/internalversion/printers.go

```diff
@@ -832,7 +832,12 @@ func printPod(pod *api.Pod, options printers.GenerateOptions) ([]metav1.TableRow
 			}
 		}
 	}
-
+	// When pod was evicted, preempted or even terminated in response to imminent node shutdown,
```

Comment:
This is probably the minimal change, but a part of me wonders whether this is about specific reasons, or about trusting the kubelet to give us an accurate reason.  The corresponding changes on the kubelet when this regressed were to make the Reason more authoritative (but it is not always so).  I wonder if we should:

1. make the reason calculated for container go to a variable called containerReason
2. have a switch statement here based on pod status reason, taking this logic and some of the existing logic that either defaults or overrides container reason
3. describe formally in pod status api reason well known reasons that supercede container reason

The problem is container reason and pod reason aren't really the same thing - the reason we return from the api is supposed to be a summary, while the container reason logic here in the printer is trying to present some useful info that would help the user in advance of the reason.

Generally, I think container reason should never be set IF pod set a reason, but we'd need to check that pod reason is that accurate in the kubelet, and then clarify in api docs.  I think that change is a bit more dramatic than this, but would help future regressions and would simplify the code (pod reason >> container reason unless unset)

Reply:
@smarterclayton Generally, a pod is set status.Reason as it becomes terminal(eg. Evicted, Rejected by admit...), And container is set state.Reason only if it is not in running state(terminal or waiting). Current, the logic in the printer is that using containerstatus.Reason to override podstatus.Reason first, but if we now use pod reason to override container reason,  this is completely reversed.  
Still, I think that keeping the original reason of containerstatus can give more details for users, but what you said still make sense to me,  so shall we only change the describe logic in printer, not change the kubelet convert api?

Reply:
> This is probably the minimal change, but a part of me wonders whether this is about specific reasons, or about trusting the kubelet to give us an accurate reason. The corresponding changes on the kubelet when this regressed were to make the Reason more authoritative (but it is not always so). I wonder if we should:
> 
> 1. make the reason calculated for container go to a variable called containerReason
> 2. have a switch statement here based on pod status reason, taking this logic and some of the existing logic that either defaults or overrides container reason
> 3. describe formally in pod status api reason well known reasons that supercede container reason
> 

we may do this in the func below https://github.com/kubernetes/kubernetes/blob/bdb9c0824c5c08d6dc7128a27e384011cfd19e64/pkg/kubelet/kubelet_pods.go#L1627


> The problem is container reason and pod reason aren't really the same thing - the reason we return from the api is supposed to be a summary, while the container reason logic here in the printer is trying to present some useful info that would help the user in advance of the reason.
> 
> Generally, I think container reason should never be set IF pod set a reason, but we'd need to check that pod reason is that accurate in the kubelet, and then clarify in api docs. I think that change is a bit more dramatic than this, but would help future regressions and would simplify the code (pod reason >> container reason unless unset)